### PR TITLE
[opt](exchange) Add buffer size limit for exchange sink operator

### PR DESCRIPTION
**Pull request by mrhhsg**
## Proposed changes

Issue Number: close #xxx

### CODE REVIEW

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

**Comment by mrhhsg** [comment]
run buildall

File: be/src/pipeline/exec/exchange_sink_buffer.cpp [hunk 1]
```diff
@@ -50,35 +50,40 @@ namespace vectorized {
...
-    }
-    std::shared_ptr<BroadcastPBlockHolder> res = _holders.top();
-    _holders.pop();
-    if (_holders.empty() && _broadcast_dependency != nullptr) {
-        _broadcast_dependency->block();
+    if (_broadcast_dependency) {
+        auto size = holder._pblock->column_values().size();
+        _total_queue_buffer_size -= size;
+        _total_queue_blocks_count--;
+        if (_total_queue_buffer_size <= 0) {
```
Comment (yiguolei) [inline be/src/pipeline/exec/exchange_sink_buffer.cpp:81]: _total_queue_buffer_size < config::exchg_node_buffer_size_bytes && _total_queue_blocks_count < config::num_broadcast_buffer?

File: be/src/pipeline/exec/exchange_sink_buffer.cpp [hunk 2]
```diff
@@ -50,35 +50,40 @@ namespace vectorized {
...
+    std::shared_ptr<BroadcastPBlockHolderMemLimiter> limiter = _parent_creator.lock();
+    if (limiter != nullptr) {
+        limiter->remove(*this);
     }
     // If the queue already deconstruted, then release pblock automatically since it
     // is a unique ptr.
 }
 
-void BroadcastPBlockHolderQueue::push(std::shared_ptr<BroadcastPBlockHolder> holder) {
+void BroadcastPBlockHolderMemLimiter::add(BroadcastPBlockHolder& holder) {
```
Comment (yiguolei) [inline be/src/pipeline/exec/exchange_sink_buffer.cpp:61]: maybe rename this method to acquire and rename remove method to release?

**Comment by mrhhsg** [comment]
run buildall

**Review by yiguolei — DISMISSED** [review]

File: be/src/pipeline/exec/exchange_sink_buffer.cpp [hunk 3]
```diff
@@ -50,35 +50,40 @@ namespace vectorized {
...
+void BroadcastPBlockHolderMemLimiter::release(const BroadcastPBlockHolder& holder) {
     std::unique_lock l(_holders_lock);
-    if (_holders.empty()) {
-        return {};
-    }
-    std::shared_ptr<BroadcastPBlockHolder> res = _holders.top();
-    _holders.pop();
-    if (_holders.empty() && _broadcast_dependency != nullptr) {
-        _broadcast_dependency->block();
+    if (_broadcast_dependency) {
```
Comment (Gabriel39) [inline be/src/pipeline/exec/exchange_sink_buffer.cpp:77]: `_broadcast_dependency` never should be nullptr

**Comment by mrhhsg** [comment]
run buildall

**Review by Gabriel39 — APPROVED** [review]

**Review by jacktengg — APPROVED** [review]

**Pull request merged**